### ISSUE TITLE
[BUG] Segmentation fault

### ISSUE THREAD

**Issue by burnmaniac** [mydumper/mydumper#1213]
**Describe the bug**
As soon as I run `myloader`, it fails with: `[1] 16008 segmentation fault  myloader --directory=db --host=127.0.0.1 --port=15003 --user=root`

**To Reproduce**
Command executed:
* `mydumper -h host -u user -P port -a -B database_name -o db -t 8 --verbose 3`
* `myloader --directory=db --host=host --port=port --user=user --password=password --verbose=3`

What mydumper and myloader version has been used?
`mydumper0.14.5-3, built against MySQL 8.0.33 with SSL support with GZIP`

**Expected behavior**
I would expect `myloader` to import the previously taken dump with `mydumper` correctly.

**Log**
```
** Message: 14:06:25.659: Using no configuration file
** Message: 14:06:25.708: Server version reported as: 5.7.41
** Message: 14:06:25.712: Initializing initialize_worker_schema
** Message: 14:06:25.811: S-Thread 1: Starting import
** Message: 14:06:25.811: S-Thread 3: Starting import
** Message: 14:06:25.812: Intermediate queue: Sending END job
** Message: 14:06:25.812: Thread 1: restoring create database on `database` from database-schema-create.sql.gz. Tables 0 of 0 completed
** Message: 14:06:25.812: Filename to open: /Users/user/db/database-schema-create.sql.gz
[1]    16008 segmentation fault  myloader --directory=db --host=127.0.0.1 --port=3306 --user=root
```

**Backup**
I can not do such thing at the moment, and I think it won't help because of the reason listed below.

**How to repeat**
For me, it fails always, doesn't matter if I dump all the tables or just one table. As soon as I start the `myloader`, it crashes.

**Environment (please complete the following information):**
 - MacBook Pro, M2 Pro
 - OS version: macOS 13.4.1
 - MyDumper version: mydumper0.14.5-3, built against MySQL 8.0.33 with SSL support with GZIP

**Additional context**
I've managed to run `lldb` and it shows that `myloader` is calling `libglib-2.0.0.dylib`g_hash_table_insert_internal` with `NULL` as first parameter.

```
(lldb) run
Process 15396 launched: '/opt/homebrew/bin/myloader' (arm64)
** Message: 13:04:52.430: Using no configuration file
** Message: 13:04:52.478: Server version reported as: 5.7.41
** Message: 13:04:52.480: Initializing initialize_worker_schema
** Message: 13:04:52.603: S-Thread 4: Starting import
** Message: 13:04:52.603: Intermediate queue: Sending END job
** Message: 13:04:52.603: Thread 4: restoring create database on `database` from database-schema-create.sql. Tables 0 of 1 completed
** Message: 13:04:52.603: Filename to open: /Users/user/db/database-schema-create.sql
** Message: 13:04:52.604: S-Thread 1: Starting import
Process 15396 stopped
* thread #12, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00000001001c17b0 libglib-2.0.0.dylib`g_str_hash
libglib-2.0.0.dylib`g_str_hash:
->  0x1001c17b0 <+0>:  ldrb   w8, [x0]
    0x1001c17b4 <+4>:  cbz    w8, 0x1001c17d4           ; <+36>
    0x1001c17b8 <+8>:  add    x9, x0, #0x1
    0x1001c17bc <+12>: mov    w0, #0x1505
Target 0: (myloader) stopped.

(lldb) f 0
frame #0: 0x00000001001c17b0 libglib-2.0.0.dylib`g_str_hash
libglib-2.0.0.dylib`g_str_hash:
->  0x1001c17b0 <+0>:  ldrb   w8, [x0]
    0x1001c17b4 <+4>:  cbz    w8, 0x1001c17d4           ; <+36>
    0x1001c17b8 <+8>:  add    x9, x0, #0x1
    0x1001c17bc <+12>: mov    w0, #0x1505

(lldb) bt
* thread #12, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x00000001001c17b0 libglib-2.0.0.dylib`g_str_hash
    frame #1: 0x00000001001c0410 libglib-2.0.0.dylib`g_hash_table_insert_internal + 60
    frame #2: 0x0000000100007678 myloader`load_schema + 436
    frame #3: 0x000000010000834c myloader`process_table_filename + 196
    frame #4: 0x000000010000e674 myloader`process_filename + 276
    frame #5: 0x000000010000e994 myloader`process_stream_filename + 28
    frame #6: 0x000000010000ea78 myloader`intermediate_thread + 64
    frame #7: 0x00000001001f8220 libglib-2.0.0.dylib`g_thread_proxy + 68
    frame #8: 0x0000000189c0bfa8 libsystem_pthread.dylib`_pthread_start + 148
```

In another situation, I've seen that the `x0` in register was all zeroes, basically `0x0000000000000000`.
I guess I don't have to provide the coredump since I've got the stack.

I've also tried running some of the previous versions, but I'm not sure which one.

I also could see that there's 0.15.* version, but I can't install it on macOS with brew, as it's not available on brew.

Let me know what else I can provide to help you with this one.

**Comment by davidducos** [comment]
Hi @burnmaniac,
The piece of code that is failing is:
```
          if ( g_str_has_prefix(dbt->table,"mydumper_")){
            g_hash_table_insert(tbl_hash, dbt->table, dbt->real_table);
          }else{
            g_hash_table_insert(tbl_hash, dbt->real_table, dbt->real_table);
```
The first parameter in both cases is tbl_hash which is initialized. There might be another issue here that we are not able to see with the current provided information. 
Is it possible for you to go to frame 2 and print the content of filename, dbt->table and dbt->real_table ?
```
(lldb) f 3
(lldb) p filename
(lldb) p dbt->table
(lldb) p dbt->real_table
```
Thanks!

**Comment by burnmaniac** [comment]
Hey @davidducos,

Thanks for bearing with me on this. I gave your suggestions a whirl, but I'm still hitting some snags. When I tried to dig into the `filename` variable in frames 2 and 3, I got:

```
(lldb) p filename
error: expression failed to parse:
error: <user expression 8>:1:1: 'filename' has unknown type; cast it to its declared type to use it
filename
^~~~~~~~
```

I even tried casting it, but then this popped up:

```
(lldb) p (int)filename
error: expression failed to parse:
error: <user expression 9>:1:6: function 'filename' with unknown type must be given a function type
(int)filename
     ^~~~~~~~
```

I also gave it a go with the `dbt` variable and its properties, but kept running into:

```
(lldb) p dbt->table
error: expression failed to parse:
error: <user expression 11>:1:1: use of undeclared identifier 'dbt'
dbt->table
^
```

Not quite sure where I'm going wrong, or if there might be something funky with my debug setup. Any other hints or tricks up your sleeve? I'm all ears and ready to get this sorted.

**Comment by davidducos** [comment]
I never used lldb, but can you try
```
(lldb) f 3
(lldb) p (char *)filename
(lldb) p (char *)(((struct db_table *)dbt)->table)
(lldb) p (char *)(((struct db_table *)dbt)->real_table)
```
?? I'm just guessing but it might work.

**Comment by burnmaniac** [comment]
Okay, I've tried it once again in `lldb` with no success, it keeps crying, so I've used the Dockerfile from this repo to build the image with debug enabled, like this:

```
docker build --build-arg CMAKE_ARGS='-DWITH_ZSTD=ON -DCMAKE_BUILD_TYPE=Debug' -t mydumper github.com/mydumper/mydumper
```

I, then, started the container, like this:

```
docker run -it --network=mynetwork --volume ~/Development/db:/db mydumper bash
```

I, then, installed the `gdb` in the container, like this:

```
yum install -y gdb
```

And, then, I run this:

```
gdb --args myloader --directory=/db --host=database --port=3306 --user=root --password=password
```

Then, I used `run`, and this is the full log:

```
[root@2ea24550d734 /]# gdb --args myloader --directory=/db --host=database --port=3306 --user=root --password=password 
GNU gdb (GDB) Red Hat Enterprise Linux 10.2-10.el9
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from myloader...
(gdb) run
Starting program: /usr/local/bin/myloader --directory=/db --host=database --port=3306 --user=root --password=password
warning: Error disabling address space randomization: Operation not permitted
Downloading separate debug info for /lib64/libz.so.1...
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[New Thread 0xffffb6f370e0 (LWP 92)]
[New Thread 0xffffb67270e0 (LWP 93)]
[New Thread 0xffffb5f170e0 (LWP 94)]
[New Thread 0xffffb57070e0 (LWP 95)]
[New Thread 0xffffb4ef70e0 (LWP 96)]
[New Thread 0xffff9ffff0e0 (LWP 97)]
[New Thread 0xffff9f7ef0e0 (LWP 98)]
[New Thread 0xffff9efdf0e0 (LWP 99)]
[New Thread 0xffff9e7cf0e0 (LWP 100)]
[New Thread 0xffff9dfbf0e0 (LWP 101)]
[New Thread 0xffff9d7af0e0 (LWP 102)]
[New Thread 0xffff9cf9f0e0 (LWP 103)]
[New Thread 0xffff87fff0e0 (LWP 104)]
[New Thread 0xffff877ef0e0 (LWP 105)]
[New Thread 0xffff86fdf0e0 (LWP 106)]
[New Thread 0xffff867cf0e0 (LWP 107)]

Thread 12 "myloader" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xffff9d7af0e0 (LWP 102)]
0x0000ffffb74ace40 in g_str_hash () from /lib64/libglib-2.0.so.0
Missing separate debuginfos, use: dnf debuginfo-install glib2-2.68.4-6.el9.aarch64 glibc-2.34-60.el9.aarch64 libgcc-11.3.1-4.3.el9.alma.aarch64 libstdc++-11.3.1-4.3.el9.alma.aarch64 mysql-community-libs-8.0.34-1.el9.aarch64 openssl-libs-3.0.7-16.el9_2.aarch64 pcre-8.44-3.el9.3.aarch64
(gdb) bt
#0  0x0000ffffb74ace40 in g_str_hash () from /lib64/libglib-2.0.so.0
#1  0x0000ffffb74b8aa8 in g_hash_table_insert () from /lib64/libglib-2.0.so.0
#2  0x000000000040e8c4 in load_schema (dbt=dbt@entry=0xffff80001700, filename=0xffff80003010 "//db/test.p_o_v-schema.sql") at /usr/src/src/myloader_process.c:260
#3  0x000000000040f040 in process_table_filename (filename=filename@entry=0xc646d40 "db.p_o_v-schema.sql") at /usr/src/src/myloader_process.c:455
#4  0x0000000000414d14 in process_filename (filename=0xc646d40 "db.p_o_v-schema.sql") at /usr/src/src/myloader_intermediate_queue.c:124
#5  process_stream_filename (iflnm=<optimized out>) at /usr/src/src/myloader_intermediate_queue.c:203
#6  0x0000000000414fec in intermediate_thread () at /usr/src/src/myloader_intermediate_queue.c:240
#7  0x0000ffffb74f5d4c in g_thread_proxy () from /lib64/libglib-2.0.so.0
#8  0x0000ffffb7043a28 in start_thread () from /lib64/libc.so.6
#9  0x0000ffffb6fecb9c in thread_start () from /lib64/libc.so.6
(gdb) f 3
#3  0x000000000040f040 in process_table_filename (filename=filename@entry=0xc646d40 "db.p_o_v-schema.sql") at /usr/src/src/myloader_process.c:455
455	/usr/src/src/myloader_process.c: Directory not empty.
(gdb) p filename
$1 = 0xc646d40 "db.p_o_v-schema.sql"
(gdb) p dbt->table
$2 = 0xffff80001c20 "p_o_v"
(gdb) p dbt->real_table
$3 = 0x0
```

I am still not sure why `lldb` is acting up and doesn't get anything useful, even after type casting, but I've found a way to reproduce it while in `gdb`. And I'd say that the reason why it fails is that the `dbt->real_table` is empty.

I hope this is enough to understand why this keeps happening. Thank you for help!

**Comment by burnmaniac** [comment]
After a bit more fiddling, I've found that the format in the dump file for the schema is `"` while `p identifier_quote_character_str` shows `` ` ``, so it fails to extract the filename.

Schema (without the fields):

```
/*!40101 SET NAMES binary*/;
/*!40014 SET FOREIGN_KEY_CHECKS=0*/;
/*!40103 SET TIME_ZONE='+00:00' */;
CREATE TABLE "p_o_v" ();
```

All other schema files from the dump have the same format.

I'm not 100% sure that this is the problem, and how can it be, since I took the dump with mydumper.

**Comment by davidducos** [comment]
Ah! then you should use --identifier-quote-character=DOUBLE_QUOTE as the source database is writing in ANSI.

**Comment by burnmaniac** [comment]
Yeah, it would be awesome if this could be detected automatically, so it can read and automatically switch between backtick and double quote, but failing with a clear message works too.

Thanks for the help!

**Comment by davidducos** [comment]
@burnmaniac 
it is very difficult for myloader to detect those kind of scenarios, not even mydumper is detecting it. That is something that you as DBA, have to take care.
So, on mydumper you also have --identifier-quote-character, as you are using ANSI, you should set that to DOBLE_QUOTE. 

if you are going to said "Hey David, but why don't we detect it automatically on mydumper??" 
my answer is simple, we give options to user to change/configure as you want, setting defaults or automatically changing might have unexpected results. Because we don't know if you want backtick and doble quotes, and both could be use with the right settings.

**Comment by burnmaniac** [comment]
I'm not a DBA, but a developer, and it wasn't me who chose ANSI compatibility, I wasn't even aware that it is enabled. Thank you for the insight.

Do you think that mydumper should also fail with a message if `--identifier-quote-character` is set to BACKTICK, instead of producing a dump that has double quotes?

**Comment by davidducos** [comment]
Would be nice to detect the mix of identifier quote character on mydumper. So, with both fix, we should not have issues like this in the future... thanks for your feedback!

**Comment by burnmaniac** [comment]
No problem! I've been trying the mydumper for a few weeks, and keep hitting issues, and I really need it as the mydump is super slow, while the database I'm working with is ~7 GB.

I do have one more question, which I think is still related to ANSI, but I can't figure out if it's me doing something wrong or is it yet another bug.

This is the command I used to take the dump:
```
mydumper -h host -u user -P 3306 -a -B db -o db --verbose 3 --identifier-quote-character DOUBLE_QUOTE
```

Once I tried to import it in a local MySQL (running in a docker container), I've got:

```
** (myloader:50805): CRITICAL **: 16:07:57.679: Thread 4: Error restoring: Unknown column 'Promo' in 'field list'
```

As far as I can see, the INSERT statement looks like:

```
INSERT INTO "type" VALUES(1,"Promo")
```

Which is invalid when ANSI is enabled. The double quotes needs to be replaced with single quotes, then it works.

Should I open another issue? Or am I missing something?
